File: app/src/main/java/org/wikipedia/search/SearchResult.java
package org.wikipedia.search;

import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.wikipedia.dataclient.WikiSite;
import org.wikipedia.dataclient.mwapi.MwQueryPage;
import org.wikipedia.model.BaseModel;
import org.wikipedia.page.PageTitle;

public class SearchResult extends BaseModel implements Parcelable {
    private final PageTitle pageTitle;
    private final String redirectFrom;
    private final SearchResultType searchResultType;

    public enum SearchResultType {
        SEARCH,
        HISTORY,
        READING_LIST,
        TAB_LIST
    }

    public SearchResult(@NonNull MwQueryPage page, @NonNull WikiSite wiki) {
        this(new PageTitle(page.title(), wiki, page.thumbUrl(), page.description(), page.displayTitle(wiki.languageCode())), page.redirectFrom(), SearchResultType.SEARCH);
    }

    public SearchResult(@NonNull PageTitle pageTitle, @NonNull SearchResultType searchResultType) {
        this(pageTitle, null, searchResultType);
    }

    public SearchResult(@NonNull PageTitle pageTitle, @Nullable String redirectFrom, @NonNull SearchResultType searchResultType) {
        this.pageTitle = pageTitle;
        this.redirectFrom = redirectFrom;
        this.searchResultType = searchResultType;
    }

    public SearchResultType getType() {
        return searchResultType;
    }

    @NonNull
    public PageTitle getPageTitle() {
        return pageTitle;
    }

    @Nullable
    public String getRedirectFrom() {
        return redirectFrom;
    }

    @Override
    public String toString() {
        return pageTitle.getPrefixedText();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Parcelable.Creator<SearchResult> CREATOR
            = new Parcelable.Creator<SearchResult>() {
        @Override
        public SearchResult createFromParcel(Parcel in) {
            return new SearchResult(in);
        }

        @Override
        public SearchResult[] newArray(int size) {
            return new SearchResult[size];
        }
    };

    @Override
    public void writeToParcel(Parcel parcel, int flags) {
        parcel.writeParcelable(pageTitle, flags);
        parcel.writeString(redirectFrom);
        parcel.writeSerializable(searchResultType);
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof SearchResult)) {
            return false;
        }
        SearchResult other = (SearchResult)o;
        return other.getPageTitle().equals(pageTitle) && TextUtils.equals(other.getRedirectFrom(), redirectFrom);
    }

    @Override
    public int hashCode() {
        int result = pageTitle.hashCode();
        result = 31 * result + redirectFrom.hashCode();
        return result;
    }

    private SearchResult(Parcel in) {
        pageTitle = in.readParcelable(PageTitle.class.getClassLoader());
        redirectFrom = in.readString();
        searchResultType = (SearchResultType) in.readSerializable();
    }
}

File: app/src/main/java/org/wikipedia/search/SearchResult.kt
package org.wikipedia.search

import android.location.Location
import kotlinx.serialization.Serializable
import org.wikipedia.dataclient.WikiSite
import org.wikipedia.dataclient.mwapi.MwQueryPage
import org.wikipedia.page.PageTitle
import org.wikipedia.util.StringUtil

@Serializable
data class SearchResult(val pageTitle: PageTitle,
                        val redirectFrom: String?,
                        val type: SearchResultType,
                        val coordinates: List<MwQueryPage.Coordinates>? = null,
                        val snippet: String? = null,
                        val indexInApiCall: Int = 0) {

    @Serializable
    enum class SearchResultType {
        PREFIX, FULL_TEXT, HISTORY, READING_LIST, TAB_LIST, SEMANTIC
    }

    constructor(
        page: MwQueryPage,
        wiki: WikiSite,
        coordinates: List<MwQueryPage.Coordinates>? = null,
        type: SearchResultType = SearchResultType.PREFIX,
        indexInApiCall: Int = 0
    ) : this(
        pageTitle = PageTitle(
            text = page.title,
            wiki = wiki,
            thumbUrl = page.thumbUrl(),
            description = page.description,
            displayText = page.displayTitle(wiki.languageCode),
        ).also { if (!page.sectionTitle.isNullOrEmpty()) it.fragment = StringUtil.addUnderscores(page.sectionTitle) },
        redirectFrom = page.redirectFrom,
        type = type,
        coordinates = coordinates,
        snippet = page.snippet,
        indexInApiCall = indexInApiCall
    )

    constructor(pageTitle: PageTitle, searchResultType: SearchResultType = SearchResultType.PREFIX, snippet: String? = null, indexInApiCall: Int = 0) :
            this(pageTitle, null, searchResultType, null, snippet, indexInApiCall)

    val location: Location? get() {
        return if (coordinates.isNullOrEmpty()) null else
            Location("").also {
                it.latitude = coordinates[0].lat
                it.longitude = coordinates[0].lon
            }
    }
}
